File: Valid Parentheses - Leetcode 20.py
class Solution:
    def isValid(self, s: str) -> bool:
        hashmap = {")": "(", "}": "{", "]": "["}
        stk = []

        for c in s:
            if c not in hashmap:
                stk.append(c)
            else:
                if not stk:
                    return False
                else:
                    popped = stk.pop()
                    if popped != hashmap[c]:
                        return False

        return not stk

# Time Complexity: O(n)
# Space Complexity: O(n)

File: Valid Parentheses - Leetcode 20/Valid Parentheses - Leetcode 20.java
import java.util.Stack;

public class Solution {
    public boolean isValid(String s) {
        java.util.Map<Character, Character> hashmap = new java.util.HashMap<>();
        hashmap.put(')', '(');
        hashmap.put('}', '{');
        hashmap.put(']', '[');
        
        Stack<Character> stk = new Stack<>();

        for (char c : s.toCharArray()) {
            if (!hashmap.containsKey(c)) {
                stk.push(c);
            } else {
                if (stk.isEmpty() || stk.pop() != hashmap.get(c)) {
                    return false;
                }
            }
        }

        return stk.isEmpty();
    }
}
